R/as_model_out_df.R

```diff
@@ -0,0 +1,211 @@
+#' already exist in `tbl`, the `tbl` column name containing `output_type` data.
+#' @param output_type_id_col character string. If an `output_type_id` column does not
+#' already exist in `tbl`, the `tbl` column name containing `output_type_id` data.
+#' @param value_col character string. If an `value` column does not
+#' already exist in `tbl`, the `tbl` column name containing `value` data.
+#' @param sep character string. Character used as separator when concatenating
+#' `team_abbr` and `model_abbr` column values into a single `model_id` string. Only
+#' applicable if `model_id` column not present and `team_abbr` and `model_abbr` columns are.
+#' @param trim_to_task_ids logical. Whether to trim `tbl` to task ID columns only. Task ID
+#' columns can be specified by providing a `<hub_connection>` class object to `hub_con` or
+#' manually through `task_id_cols`.
+#' @param hub_con a `<hub_connection>` class object. Only used if
+#' `trim_to_task_ids = TRUE` and tasks IDs should be determined from the hub config.
+#' @param task_id_cols a character vector of column names. Only used if
+#' `trim_to_task_ids = TRUE` to manually specify task ID columns to retain.
+#' Overrides `hub_con` argument if provided.
+#' @param remove_empty Logical. Whether to remove columns containing only `NA`.
+#'
+#' @return A `model_out_tbl` class object.
+#' @export
+#'
+#' @examples
+#' library(dplyr)
+#' hub_path <- system.file("testhubs/flusight", package = "hubUtils")
+#' hub_con <- connect_hub(hub_path)
+#' hub_con %>%
+#'   filter(output_type == "quantile", location == "US") %>%
+#'   collect() %>%
+#'   filter(forecast_date == max(forecast_date)) %>%
+#'   as_model_out_tbl()
+#'
+as_model_out_tbl <- function(tbl, model_id_col = NULL, output_type_col = NULL,
+                             output_type_id_col = NULL, value_col = NULL, sep = "-",
+                             trim_to_task_ids = FALSE, hub_con = NULL,
+                             task_id_cols = NULL, remove_empty = FALSE) {
+  checkmate::assert_data_frame(tbl)
+  tbl <- tibble::as_tibble(tbl)
+
+  tbl <- rename_columns(tbl,
+    model_id_col = model_id_col,
+    output_type_col = output_type_col,
+    output_type_id_col = output_type_id_col,
+    value_col = value_col
+  )
+
+  if (!"model_id" %in% names(tbl)) {
+    cli::cli_alert_warning("{.arg model_id} column missing. Attempting to create automatically.")
+    tbl <- model_id_merge(tbl, sep = sep)
+  }
+
+  if (trim_to_task_ids) {
+    tbl <- trim_tbl_to_task_ids(
+      tbl = tbl,
+      task_id_cols = task_id_cols,
+      hub_con = hub_con
+    )
+  }
+
+  if (remove_empty) {
+    tbl <- remove_model_out_tbl_empty(tbl)
+  }
+
+  tbl <- std_col_order_model_out_tbl(tbl) %>%
+    std_row_order_model_out_tbl()
+
+  validate_model_out_tbl(tbl)
+
+  structure(tbl,
+    class = c("model_out_tbl", class(tbl))
+  )
+}
+
+# --- Unexported Utilities ----
+validate_model_out_tbl <- function(tbl) {
```

Comment:
Might we want also to check for the correct datatypes in columns that have unambiguous datatypes?

Reply:
Do you mean checking against the schema generated from a tasks.json config? It would make a `hub_connection()` an additional required argument to the validator and therefore `as_model_out_tbl()` too.

Reply:
Another interesting question is what to do about a potentially ambiguous column like `output_type_id`. Say the hub contains some pmf data so is a character column but we've extracted only cdf data that can be cast to numeric. Should this ever be done and would `as_model_out_tbl()` be the place to do it?

Thinking about it this might be complicated and not the right place to do this and likely better handled within ensembling or plotting functions but useful to think about.

Reply:
isn't there an ironclad rule that `value` is numeric and `output_type` is character? if so, maybe worth checking those? again, not a strongly held preference here.

Reply:
It does seem like at minimum, we should be able to check that `model_id` and `output_type` are character type and `value` is numeric?

I think I agree that dealing with the type of `output_type_id` seems complicated and is maybe best handled elsewhere...

Reply:
OK, I've implemented basic datatype checks in :

Currently the function will error if `tbl` does not contain the correct datatypes.

One question I had is about factors. One may well have converted  model_id or output_type to factor (for example for plotting in a specific order). Factors will likely work as character in most downstream cases I can think of (e.g. summarise reprex) below but will currently error.

Is this what we want?

``` r
library(hubUtils)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

hub_con <- connect_hub(system.file("testhubs/flusight", package = "hubUtils"))
tbl <- hub_con %>%
    dplyr::filter(output_type == "quantile", location == "US") %>%
    dplyr::collect() %>%
    dplyr::filter(forecast_date == max(forecast_date)) %>%
    dplyr::arrange(model_id)

group_by(tbl, model_id) %>% summarise(mean = mean(value))
#> # A tibble: 2 × 2
#>   model_id      mean
#>   <chr>        <dbl>
#> 1 hub-baseline 1513.
#> 2 hub-ensemble 1018.


tbl$model_id <- as.factor(tbl$model_id)
group_by(tbl, model_id) %>% summarise(mean = mean(value))
#> # A tibble: 2 × 2
#>   model_id      mean
#>   <fct>        <dbl>
#> 1 hub-baseline 1513.
#> 2 hub-ensemble 1018.
```

<sup>Created on 2023-06-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
I don't have a strong feeling about what the "right" way to do this is. If it's not that complicated to do so, it feels like it keeps it maximally flexible to allow for those two columns to be factors as well. But I'm also ok being strict about it and demanding that they be characters.

Reply:
It's not too complicated so happy to implement. Would prefer to get a thumbs up from @elray1 too before I do so. Any feelings about factors @elray?

Reply:
I've gone ahead and implemented allowing factors to be valid data types for `model_id` and `output_type` in 7ac1b9c16afd21419ce02189cd4e0fe02a36b3b5. Also fixed a small bug. Do you mind approving this PR @nickreich so we can merge and ficus on #86 ? We can easily revert if necessary.